Multi-dimensional array input type causes a compilation error

So I have the following schema:
```
type Schedule {
    ...
    occurrences: [[String]]
}
```
```
input ScheduleInput {
    ...
    occurrences: [[String]]
}
```
I have the following query:

```
mutation AddActivity($schedule : ScheduleInput) {
        addActivity(schedule: $schedule) {
             id
        }
}
```

A lot of fields have been left out for simplification as they don't matter.

I'm using this schema & query on Amazon AppSync. It's validated and it is working. I'm now trying to hook up an Android sample app to the AppSync instance running this GraphQL schema, using the apollo-android library.

The project builds fine at first, all the classes/queries/mutations are being generated fine, but there's some sort of compilation error when I actually try and run the app. **If I remove `occurrences : [[String]]` from the schema, the error is gone and it runs fine.**

Please find a complete stack-trace of the problem at the bottom of this comment. The actual error is the following:

```
/Users/andreiharasemiuc/GraphQLSample/app/build/generated/source/apollo/apollotest/mq/apollotest/api/type/ScheduleInput.java:79: error: cannot find symbol
12:38:43.444 [ERROR] [system.err]                 listItemWriter.writeObject($item.marshaller());
```

ScheduleInput.java is a class generated by the library. Here's the concerned code from that class:

```
  public InputFieldMarshaller marshaller() {
    return new InputFieldMarshaller() {
      @Override
      public void marshal(InputFieldWriter writer) throws IOException {
        if (type.defined) {
          writer.writeLong("type", type.value);
        }
        if (startDate.defined) {
          writer.writeString("startDate", startDate.value);
        }
        if (endDate.defined) {
          writer.writeString("endDate", endDate.value);
        }
        if (occurrences.defined) {
          writer.writeList("occurrences", occurrences.value != null ? new InputFieldWriter.ListWriter() {
            @Override
            public void write(InputFieldWriter.ListItemWriter listItemWriter) throws IOException {
              for (List<String> $item : occurrences.value) {
                listItemWriter.writeObject($item.marshaller());
              }
            }
          } : null);
        }
        if (timeUnitsToSkip.defined) {
          writer.writeLong("timeUnitsToSkip", timeUnitsToSkip.value);
        }
      }
    };
  }
```


The stack-trace:

```
Error:12:38:43.444 [ERROR] [system.err] /Users/andreiharasemiuc/GraphQLSample/app/build/generated/source/apollo/apollotest/mq/apollotest/api/type/ScheduleInput.java:79: error: cannot find symbol
12:38:43.444 [ERROR] [system.err]                 listItemWriter.writeObject($item.marshaller());
12:38:43.444 [ERROR] [system.err]                                                 ^
12:38:43.444 [ERROR] [system.err]   symbol:   method marshaller()
12:38:43.445 [ERROR] [system.err]   location: variable $item of type List<String>
12:38:43.446 [ERROR] [system.err] 1 error
12:38:43.463 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 
12:38:43.463 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] FAILURE: Build failed with an exception.
12:38:43.463 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 
12:38:43.464 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] * What went wrong:
12:38:43.464 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] Execution failed for task ':app:compileDebugJavaWithJavac'.
12:38:43.464 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] > Compilation failed; see the compiler error output for details.
12:38:43.464 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 
12:38:43.464 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] * Exception is:
12:38:43.464 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':app:compileDebugJavaWithJavac'.
12:38:43.464 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:100)
12:38:43.464 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:70)
12:38:43.464 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:63)
12:38:43.464 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
12:38:43.464 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)
12:38:43.464 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:88)
12:38:43.464 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:52)
12:38:43.465 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
12:38:43.465 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
12:38:43.465 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
12:38:43.465 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
12:38:43.465 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:248)
12:38:43.465 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
12:38:43.465 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
12:38:43.465 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:197)
12:38:43.465 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:107)
12:38:43.465 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:241)
12:38:43.465 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:230)
12:38:43.465 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.processTask(DefaultTaskPlanExecutor.java:124)
12:38:43.465 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.access$200(DefaultTaskPlanExecutor.java:80)
12:38:43.466 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:105)
12:38:43.466 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:99)
12:38:43.466 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:625)
12:38:43.466 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:580)
12:38:43.466 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:99)
12:38:43.466 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
12:38:43.466 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
12:38:43.466 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
12:38:43.466 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] Caused by: org.gradle.api.internal.tasks.compile.CompilationFailedException: Compilation failed; see the compiler error output for details.
12:38:43.466 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:51)
12:38:43.466 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:36)
12:38:43.466 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler.delegateAndHandleErrors(NormalizingJavaCompiler.java:99)
12:38:43.466 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler.execute(NormalizingJavaCompiler.java:52)
12:38:43.467 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler.execute(NormalizingJavaCompiler.java:37)
12:38:43.467 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.tasks.compile.CleaningJavaCompilerSupport.execute(CleaningJavaCompilerSupport.java:35)
12:38:43.467 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.tasks.compile.CleaningJavaCompilerSupport.execute(CleaningJavaCompilerSupport.java:25)
12:38:43.467 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.tasks.compile.incremental.IncrementalCompilationFinalizer.execute(IncrementalCompilationFinalizer.java:39)
12:38:43.467 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.tasks.compile.incremental.IncrementalCompilationFinalizer.execute(IncrementalCompilationFinalizer.java:24)
12:38:43.467 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.tasks.compile.JavaCompile.performCompilation(JavaCompile.java:198)
12:38:43.467 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.tasks.compile.JavaCompile.compile(JavaCompile.java:129)
12:38:43.467 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at com.android.build.gradle.tasks.factory.AndroidJavaCompile.compile(AndroidJavaCompile.java:95)
12:38:43.467 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
12:38:43.467 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$IncrementalTaskAction.doExecute(DefaultTaskClassInfoStore.java:173)
12:38:43.467 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:134)
12:38:43.467 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:121)
12:38:43.467 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.run(ExecuteActionsTaskExecuter.java:122)
12:38:43.467 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
12:38:43.468 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
12:38:43.468 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:197)
12:38:43.468 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:107)
12:38:43.468 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:111)
12:38:43.468 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:92)
12:38:43.468 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	... 27 more
12:38:43.468 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 
12:38:43.468 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 
12:38:43.468 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] * Get more help at https://help.gradle.org
12:38:43.468 [ERROR] [org.gradle.internal.buildevents.BuildResultLogger] 
12:38:43.468 [ERROR] [org.gradle.internal.buildevents.BuildResultLogger] BUILD FAILED in 2s
```


I'm using the following dependencies

```
buildscript {
    
    repositories {
        google()
        jcenter()
        maven { url 'https://oss.sonatype.org/content/repositories/snapshots/' }
    }
    dependencies {
        classpath 'com.android.tools.build:gradle:3.0.1'
        classpath 'com.apollographql.apollo:apollo-gradle-plugin:0.4.4-SNAPSHOT'
        

        // NOTE: Do not place your application dependencies here; they belong
        // in the individual module build.gradle files
    }
}

    compile 'com.apollographql.apollo:apollo-runtime:0.4.4-SNAPSHOT'
    compile "com.apollographql.apollo:apollo-android-support:0.4.4-SNAPSHOT"
    compile "com.apollographql.apollo:apollo-rx2-support:0.4.4-SNAPSHOT"

```

--- a/.travis.yml
+++ b/.travis.yml
@@ -8,7 +8,7 @@ android:
     - tools
     - platform-tools
     - tools  # Upgrade again after upgrading platform-tools
-    - build-tools-25.0.3
+    - build-tools-26.0.2
     - android-22
     - extra-google-m2repository
     - extra-android-m2repository

--- a/apollo-api/src/main/java/com/apollographql/apollo/api/InputFieldWriter.java
+++ b/apollo-api/src/main/java/com/apollographql/apollo/api/InputFieldWriter.java
@@ -39,5 +39,7 @@ public interface InputFieldWriter {
     void writeCustom(ScalarType scalarType, Object value) throws IOException;
 
     void writeObject(InputFieldMarshaller marshaller) throws IOException;
+
+    void writeList(ListWriter listWriter) throws IOException;
   }
 }

--- a/apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/InputFieldSpec.kt
+++ b/apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/InputFieldSpec.kt
@@ -25,9 +25,7 @@ class InputFieldSpec(
       Type.ENUM -> writeEnumCode(writerParam)
       Type.CUSTOM -> writeCustomCode(writerParam)
       Type.OBJECT -> writeObjectCode(writerParam, marshaller)
-      Type.SCALAR_LIST -> writeScalarList(writerParam)
-      Type.CUSTOM_LIST -> writeCustomList(writerParam)
-      Type.OBJECT_LIST -> writeObjectList(writerParam, marshaller)
+      Type.LIST -> writeList(writerParam, marshaller)
     }.let {
       if (javaType.isOptional()) {
         CodeBlock.builder()
@@ -68,88 +66,71 @@ class InputFieldSpec(
     return CodeBlock.of("\$L.\$L(\$S, \$L);\n", writerParam, WRITE_METHODS[type], name, valueCode)
   }
 
-  private fun writeScalarList(writerParam: CodeBlock): CodeBlock {
-    val rawFieldType = with(javaType.unwrapOptionalType(true)) { if (isList()) listParamType() else this }
-    val writeMethod = SCALAR_LIST_ITEM_WRITE_METHODS[rawFieldType] ?: "writeString"
-    val writeStatement = CodeBlock.builder()
-        .beginControlFlow("for (\$T \$L : \$L)", rawFieldType, "\$item",
-            javaType.unwrapOptionalValue(name, false))
-        .add(
-            if (rawFieldType.isEnum(context)) {
-              CodeBlock.of("\$L.\$L(\$L.name());\n", LIST_ITEM_WRITER_PARAM.name, writeMethod, "\$item")
-            } else {
-              CodeBlock.of("\$L.\$L(\$L);\n", LIST_ITEM_WRITER_PARAM.name, writeMethod, "\$item")
-            })
-        .endControlFlow()
-        .build()
-    val listWriterType = TypeSpec.anonymousClassBuilder("")
-        .addSuperinterface(InputFieldWriter.ListWriter::class.java)
-        .addMethod(MethodSpec.methodBuilder("write")
-            .addModifiers(Modifier.PUBLIC)
-            .addAnnotation(Override::class.java)
-            .addException(IOException::class.java)
-            .addParameter(LIST_ITEM_WRITER_PARAM)
-            .addCode(writeStatement)
-            .build()
-        )
-        .build()
-    val valueCode = javaType.unwrapOptionalValue(name) {
-      CodeBlock.of("\$L", listWriterType)
+  private fun writeList(writerParam: CodeBlock, marshaller: CodeBlock): CodeBlock {
+    val unwrappedListValue = javaType.unwrapOptionalValue(name, false)
+    val listParamType = with(javaType.unwrapOptionalType(true)) { if (isList()) listParamType() else this }
+    val listWriterCode = javaType.unwrapOptionalValue(name) {
+      CodeBlock.of("\$L", listWriter(listParamType, unwrappedListValue, "\$item", marshaller))
     }
-    return CodeBlock.of("\$L.\$L(\$S, \$L);\n", writerParam, WRITE_METHODS[type], name, valueCode)
+    return CodeBlock.of("\$L.\$L(\$S, \$L);\n", writerParam, WRITE_METHODS[Type.LIST], name, listWriterCode)
   }
 
-  private fun writeCustomList(writerParam: CodeBlock): CodeBlock {
-    val rawFieldType = with(javaType.unwrapOptionalType(true)) { if (isList()) listParamType() else this }
-    val customScalarEnum = CustomEnumTypeSpecBuilder.className(context)
-    val customScalarEnumConst = normalizeGraphQlType(graphQLType).toUpperCase(Locale.ENGLISH)
+  private fun listWriter(itemType: TypeName, listParam: CodeBlock, itemParam: String, marshaller: CodeBlock): TypeSpec {
     val writeStatement = CodeBlock.builder()
-        .beginControlFlow("for (\$T \$L : \$L)", rawFieldType, "\$item",
-            javaType.unwrapOptionalValue(name, false))
-        .addStatement("\$L.writeCustom(\$T.\$L, \$L)", LIST_ITEM_WRITER_PARAM.name, customScalarEnum,
-            customScalarEnumConst, "\$item")
+        .beginControlFlow("for (\$T \$L : \$L)", itemType, itemParam, listParam)
+        .add(writeListItemStatement(itemType, itemParam, marshaller))
         .endControlFlow()
         .build()
-    val listWriterType = TypeSpec.anonymousClassBuilder("")
+    return TypeSpec.anonymousClassBuilder("")
         .addSuperinterface(InputFieldWriter.ListWriter::class.java)
         .addMethod(MethodSpec.methodBuilder("write")
             .addModifiers(Modifier.PUBLIC)
             .addAnnotation(Override::class.java)
-            .addParameter(LIST_ITEM_WRITER_PARAM)
             .addException(IOException::class.java)
+            .addParameter(LIST_ITEM_WRITER_PARAM)
             .addCode(writeStatement)
             .build()
         )
         .build()
-    val valueCode = javaType.unwrapOptionalValue(name) {
-      CodeBlock.of("\$L", listWriterType)
-    }
-    return CodeBlock.of("\$L.\$L(\$S, \$L);\n", writerParam, WRITE_METHODS[type], name, valueCode)
   }
 
-  private fun writeObjectList(writerParam: CodeBlock, marshaller: CodeBlock): CodeBlock {
-    val rawFieldType = with(javaType.unwrapOptionalType(true)) { if (isList()) listParamType() else this }
-    val writeStatement = CodeBlock.builder()
-        .beginControlFlow("for (\$T \$L : \$L)", rawFieldType, "\$item",
-            javaType.unwrapOptionalValue(name, false))
-        .addStatement("\$L.writeObject(\$L.\$L)", LIST_ITEM_WRITER_PARAM.name, "\$item", marshaller)
-        .endControlFlow()
-        .build()
-    val listWriterType = TypeSpec.anonymousClassBuilder("")
-        .addSuperinterface(InputFieldWriter.ListWriter::class.java)
-        .addMethod(MethodSpec.methodBuilder("write")
-            .addModifiers(Modifier.PUBLIC)
-            .addAnnotation(Override::class.java)
-            .addParameter(LIST_ITEM_WRITER_PARAM)
-            .addException(IOException::class.java)
-            .addCode(writeStatement)
-            .build()
-        )
-        .build()
-    val valueCode = javaType.unwrapOptionalValue(name) {
-      CodeBlock.of("\$L", listWriterType)
+  private fun writeListItemStatement(itemType: TypeName, itemParam: String, marshaller: CodeBlock): CodeBlock {
+    fun writeList(): CodeBlock {
+      val nestedListItemParamType = with(itemType.unwrapOptionalType(true)) { if (isList()) listParamType() else this }
+      val nestedListWriter = listWriter(nestedListItemParamType, CodeBlock.of("\$L", itemParam), "\$$itemParam",
+          marshaller)
+      return CodeBlock.of("\$L.\$L(\$L != null ? \$L : null);\n", LIST_ITEM_WRITER_PARAM.name, WRITE_METHODS[Type.LIST],
+          itemParam, nestedListWriter)
     }
-    return CodeBlock.of("\$L.\$L(\$S, \$L);\n", writerParam, WRITE_METHODS[type], name, valueCode)
+
+    fun writeCustom(): CodeBlock {
+      val customScalarEnum = CustomEnumTypeSpecBuilder.className(context)
+      val customScalarEnumConst = normalizeGraphQlType(graphQLType, true).toUpperCase(Locale.ENGLISH)
+      return CodeBlock.of("\$L.writeCustom(\$T.\$L, \$L);\n", LIST_ITEM_WRITER_PARAM.name, customScalarEnum,
+          customScalarEnumConst, itemParam)
+    }
+
+    fun writeScalar(): CodeBlock {
+      val writeMethod = SCALAR_LIST_ITEM_WRITE_METHODS[itemType] ?: "writeString"
+      return if (itemType.isEnum(context)) {
+        CodeBlock.of("\$L.\$L(\$L != null ? \$L.name() : null);\n", LIST_ITEM_WRITER_PARAM.name, writeMethod, itemParam,
+            itemParam)
+      } else {
+        CodeBlock.of("\$L.\$L(\$L);\n", LIST_ITEM_WRITER_PARAM.name, writeMethod, itemParam)
+      }
+    }
+
+    fun writeObject(): CodeBlock {
+      return CodeBlock.of("\$L.writeObject(\$L != null ? \$L.\$L : null);\n", LIST_ITEM_WRITER_PARAM.name, itemParam,
+          itemParam, marshaller);
+    }
+
+    return when {
+      itemType.isList() -> writeList()
+      graphQLType.isCustomScalarType(context) -> return writeCustom()
+      itemType.isScalar(context) -> return writeScalar()
+      else -> return writeObject()
+    }
   }
 
   companion object {
@@ -162,9 +143,7 @@ class InputFieldSpec(
         Type.ENUM to "writeString",
         Type.CUSTOM to "writeCustom",
         Type.OBJECT to "writeObject",
-        Type.SCALAR_LIST to "writeList",
-        Type.CUSTOM_LIST to "writeList",
-        Type.OBJECT_LIST to "writeList"
+        Type.LIST to "writeList"
     )
     private val SCALAR_LIST_ITEM_WRITE_METHODS = mapOf(
         ClassNames.STRING to "writeString",
@@ -186,16 +165,7 @@ class InputFieldSpec(
           .resolve(typeName = graphQLType, nullableValueType = nullableValueType)
       val normalizedJavaType = javaType.unwrapOptionalType(true)
       val type = when {
-        normalizedJavaType.isList() -> {
-          val rawFieldType = normalizedJavaType.let { if (it.isList()) it.listParamType() else it }
-          if (graphQLType.isCustomScalarType(context)) {
-            Type.CUSTOM_LIST
-          } else if (rawFieldType.isScalar(context)) {
-            Type.SCALAR_LIST
-          } else {
-            Type.OBJECT_LIST
-          }
-        }
+        normalizedJavaType.isList() -> Type.LIST
         graphQLType.isCustomScalarType(context) -> Type.CUSTOM
         normalizedJavaType.isScalar(context) -> {
           when (normalizedJavaType) {
@@ -226,9 +196,7 @@ class InputFieldSpec(
     BOOLEAN,
     ENUM,
     OBJECT,
-    SCALAR_LIST,
-    CUSTOM_LIST,
-    OBJECT_LIST,
+    LIST,
     CUSTOM,
   }
 }

--- a/apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/Util.kt
+++ b/apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/Util.kt
@@ -393,13 +393,25 @@ fun TypeName.isList() =
 fun TypeName.isEnum(context: CodeGenerationContext) =
     ((this is ClassName) && context.typeDeclarations.count { it.kind == "EnumType" && it.name == simpleName() } > 0)
 
-fun String.isCustomScalarType(context: CodeGenerationContext) =
-    context.customTypeMap.containsKey(normalizeGraphQlType(this))
+fun String.isCustomScalarType(context: CodeGenerationContext): Boolean {
+  val normalizedType = normalizeGraphQlType(this)
+  return if (this != normalizedType) {
+    normalizedType.isCustomScalarType(context)
+  } else {
+    context.customTypeMap.containsKey(normalizedType)
+  }
+}
 
 fun TypeName.isScalar(context: CodeGenerationContext) = (Util.SCALAR_TYPES.contains(this) || isEnum(context))
 
-fun normalizeGraphQlType(type: String) =
-    type.removeSuffix("!").removeSurrounding(prefix = "[", suffix = "]").removeSuffix("!")
+fun normalizeGraphQlType(type: String, recursive: Boolean = false): String {
+  val normalizedType = type.removeSuffix("!").removeSurrounding(prefix = "[", suffix = "]").removeSuffix("!")
+  return if (recursive && normalizedType != type) {
+    normalizeGraphQlType(normalizedType, true)
+  } else {
+    normalizedType
+  }
+}
 
 fun TypeName.listParamType(): TypeName {
   return (this as ParameterizedTypeName)

--- a/apollo-compiler/src/test/graphql/com/example/input_object_type/TestOperation.json
+++ b/apollo-compiler/src/test/graphql/com/example/input_object_type/TestOperation.json
@@ -198,6 +198,26 @@
 					"type": "Boolean",
 					"description": "for test purpose only",
 					"defaultValue": true
+				},
+				{
+					"name": "listOfListOfString",
+					"type": "[[String]]",
+					"description": "for test purpose only"
+				},
+				{
+					"name": "listOfListOfEnum",
+					"type": "[[Episode]]",
+					"description": "for test purpose only"
+				},
+				{
+					"name": "listOfListOfCustom",
+					"type": "[[Date]]",
+					"description": "for test purpose only"
+				},
+				{
+					"name": "listOfListOfObject",
+					"type": "[[ColorInput]]",
+					"description": "for test purpose only"
 				}
 			]
 		},

--- a/apollo-compiler/src/test/graphql/com/example/input_object_type/type/ReviewInput.java
+++ b/apollo-compiler/src/test/graphql/com/example/input_object_type/type/ReviewInput.java
@@ -42,11 +42,22 @@ public final class ReviewInput {
 
   private final Input<Boolean> booleanWithDefaultValue;
 
+  private final Input<List<List<String>>> listOfListOfString;
+
+  private final Input<List<List<Episode>>> listOfListOfEnum;
+
+  private final Input<List<List<Date>>> listOfListOfCustom;
+
+  private final Input<List<List<ColorInput>>> listOfListOfObject;
+
   ReviewInput(long stars, Input<Long> nullableIntFieldWithDefaultValue, Input<String> commentary,
       @Nonnull ColorInput favoriteColor, Input<Episode> enumWithDefaultValue,
       Input<Episode> nullableEnum, Input<List<Date>> listOfCustomScalar, Input<Date> customScalar,
       Input<List<Episode>> listOfEnums, Input<List<Long>> listOfInt,
-      Input<List<String>> listOfString, Input<Boolean> booleanWithDefaultValue) {
+      Input<List<String>> listOfString, Input<Boolean> booleanWithDefaultValue,
+      Input<List<List<String>>> listOfListOfString, Input<List<List<Episode>>> listOfListOfEnum,
+      Input<List<List<Date>>> listOfListOfCustom,
+      Input<List<List<ColorInput>>> listOfListOfObject) {
     this.stars = stars;
     this.nullableIntFieldWithDefaultValue = nullableIntFieldWithDefaultValue;
     this.commentary = commentary;
@@ -59,6 +70,10 @@ public final class ReviewInput {
     this.listOfInt = listOfInt;
     this.listOfString = listOfString;
     this.booleanWithDefaultValue = booleanWithDefaultValue;
+    this.listOfListOfString = listOfListOfString;
+    this.listOfListOfEnum = listOfListOfEnum;
+    this.listOfListOfCustom = listOfListOfCustom;
+    this.listOfListOfObject = listOfListOfObject;
   }
 
   /**
@@ -145,6 +160,34 @@ public final class ReviewInput {
     return this.booleanWithDefaultValue.value;
   }
 
+  /**
+   * for test purpose only
+   */
+  public @Nullable List<List<String>> listOfListOfString() {
+    return this.listOfListOfString.value;
+  }
+
+  /**
+   * for test purpose only
+   */
+  public @Nullable List<List<Episode>> listOfListOfEnum() {
+    return this.listOfListOfEnum.value;
+  }
+
+  /**
+   * for test purpose only
+   */
+  public @Nullable List<List<Date>> listOfListOfCustom() {
+    return this.listOfListOfCustom.value;
+  }
+
+  /**
+   * for test purpose only
+   */
+  public @Nullable List<List<ColorInput>> listOfListOfObject() {
+    return this.listOfListOfObject.value;
+  }
+
   public static Builder builder() {
     return new Builder();
   }
@@ -185,7 +228,7 @@ public final class ReviewInput {
             @Override
             public void write(InputFieldWriter.ListItemWriter listItemWriter) throws IOException {
               for (Episode $item : listOfEnums.value) {
-                listItemWriter.writeString($item.name());
+                listItemWriter.writeString($item != null ? $item.name() : null);
               }
             }
           } : null);
@@ -212,6 +255,78 @@ public final class ReviewInput {
         }
         if (booleanWithDefaultValue.defined) {
           writer.writeBoolean("booleanWithDefaultValue", booleanWithDefaultValue.value);
+        }
+        if (listOfListOfString.defined) {
+          writer.writeList("listOfListOfString", listOfListOfString.value != null ? new InputFieldWriter.ListWriter() {
+            @Override
+            public void write(InputFieldWriter.ListItemWriter listItemWriter) throws IOException {
+              for (List<String> $item : listOfListOfString.value) {
+                listItemWriter.writeList($item != null ? new InputFieldWriter.ListWriter() {
+                  @Override
+                  public void write(InputFieldWriter.ListItemWriter listItemWriter) throws
+                      IOException {
+                    for (String $$item : $item) {
+                      listItemWriter.writeString($$item);
+                    }
+                  }
+                } : null);
+              }
+            }
+          } : null);
+        }
+        if (listOfListOfEnum.defined) {
+          writer.writeList("listOfListOfEnum", listOfListOfEnum.value != null ? new InputFieldWriter.ListWriter() {
+            @Override
+            public void write(InputFieldWriter.ListItemWriter listItemWriter) throws IOException {
+              for (List<Episode> $item : listOfListOfEnum.value) {
+                listItemWriter.writeList($item != null ? new InputFieldWriter.ListWriter() {
+                  @Override
+                  public void write(InputFieldWriter.ListItemWriter listItemWriter) throws
+                      IOException {
+                    for (Episode $$item : $item) {
+                      listItemWriter.writeString($$item != null ? $$item.name() : null);
+                    }
+                  }
+                } : null);
+              }
+            }
+          } : null);
+        }
+        if (listOfListOfCustom.defined) {
+          writer.writeList("listOfListOfCustom", listOfListOfCustom.value != null ? new InputFieldWriter.ListWriter() {
+            @Override
+            public void write(InputFieldWriter.ListItemWriter listItemWriter) throws IOException {
+              for (List<Date> $item : listOfListOfCustom.value) {
+                listItemWriter.writeList($item != null ? new InputFieldWriter.ListWriter() {
+                  @Override
+                  public void write(InputFieldWriter.ListItemWriter listItemWriter) throws
+                      IOException {
+                    for (Date $$item : $item) {
+                      listItemWriter.writeCustom(CustomType.DATE, $$item);
+                    }
+                  }
+                } : null);
+              }
+            }
+          } : null);
+        }
+        if (listOfListOfObject.defined) {
+          writer.writeList("listOfListOfObject", listOfListOfObject.value != null ? new InputFieldWriter.ListWriter() {
+            @Override
+            public void write(InputFieldWriter.ListItemWriter listItemWriter) throws IOException {
+              for (List<ColorInput> $item : listOfListOfObject.value) {
+                listItemWriter.writeList($item != null ? new InputFieldWriter.ListWriter() {
+                  @Override
+                  public void write(InputFieldWriter.ListItemWriter listItemWriter) throws
+                      IOException {
+                    for (ColorInput $$item : $item) {
+                      listItemWriter.writeObject($$item != null ? $$item.marshaller() : null);
+                    }
+                  }
+                } : null);
+              }
+            }
+          } : null);
         }
       }
     };
@@ -241,6 +356,14 @@ public final class ReviewInput {
     private Input<List<String>> listOfString = Input.fromNullable(Arrays.<String>asList("test1", "test2", "test3"));
 
     private Input<Boolean> booleanWithDefaultValue = Input.fromNullable(true);
+
+    private Input<List<List<String>>> listOfListOfString = Input.absent();
+
+    private Input<List<List<Episode>>> listOfListOfEnum = Input.absent();
+
+    private Input<List<List<Date>>> listOfListOfCustom = Input.absent();
+
+    private Input<List<List<ColorInput>>> listOfListOfObject = Input.absent();
 
     Builder() {
     }
@@ -344,6 +467,38 @@ public final class ReviewInput {
     /**
      * for test purpose only
      */
+    public Builder listOfListOfString(@Nullable List<List<String>> listOfListOfString) {
+      this.listOfListOfString = Input.fromNullable(listOfListOfString);
+      return this;
+    }
+
+    /**
+     * for test purpose only
+     */
+    public Builder listOfListOfEnum(@Nullable List<List<Episode>> listOfListOfEnum) {
+      this.listOfListOfEnum = Input.fromNullable(listOfListOfEnum);
+      return this;
+    }
+
+    /**
+     * for test purpose only
+     */
+    public Builder listOfListOfCustom(@Nullable List<List<Date>> listOfListOfCustom) {
+      this.listOfListOfCustom = Input.fromNullable(listOfListOfCustom);
+      return this;
+    }
+
+    /**
+     * for test purpose only
+     */
+    public Builder listOfListOfObject(@Nullable List<List<ColorInput>> listOfListOfObject) {
+      this.listOfListOfObject = Input.fromNullable(listOfListOfObject);
+      return this;
+    }
+
+    /**
+     * for test purpose only
+     */
     public Builder nullableIntFieldWithDefaultValueInput(@Nonnull Input<Long> nullableIntFieldWithDefaultValue) {
       this.nullableIntFieldWithDefaultValue = Utils.checkNotNull(nullableIntFieldWithDefaultValue, "nullableIntFieldWithDefaultValue == null");
       return this;
@@ -421,9 +576,41 @@ public final class ReviewInput {
       return this;
     }
 
+    /**
+     * for test purpose only
+     */
+    public Builder listOfListOfStringInput(@Nonnull Input<List<List<String>>> listOfListOfString) {
+      this.listOfListOfString = Utils.checkNotNull(listOfListOfString, "listOfListOfString == null");
+      return this;
+    }
+
+    /**
+     * for test purpose only
+     */
+    public Builder listOfListOfEnumInput(@Nonnull Input<List<List<Episode>>> listOfListOfEnum) {
+      this.listOfListOfEnum = Utils.checkNotNull(listOfListOfEnum, "listOfListOfEnum == null");
+      return this;
+    }
+
+    /**
+     * for test purpose only
+     */
+    public Builder listOfListOfCustomInput(@Nonnull Input<List<List<Date>>> listOfListOfCustom) {
+      this.listOfListOfCustom = Utils.checkNotNull(listOfListOfCustom, "listOfListOfCustom == null");
+      return this;
+    }
+
+    /**
+     * for test purpose only
+     */
+    public Builder listOfListOfObjectInput(@Nonnull Input<List<List<ColorInput>>> listOfListOfObject) {
+      this.listOfListOfObject = Utils.checkNotNull(listOfListOfObject, "listOfListOfObject == null");
+      return this;
+    }
+
     public ReviewInput build() {
       Utils.checkNotNull(favoriteColor, "favoriteColor == null");
-      return new ReviewInput(stars, nullableIntFieldWithDefaultValue, commentary, favoriteColor, enumWithDefaultValue, nullableEnum, listOfCustomScalar, customScalar, listOfEnums, listOfInt, listOfString, booleanWithDefaultValue);
+      return new ReviewInput(stars, nullableIntFieldWithDefaultValue, commentary, favoriteColor, enumWithDefaultValue, nullableEnum, listOfCustomScalar, customScalar, listOfEnums, listOfInt, listOfString, booleanWithDefaultValue, listOfListOfString, listOfListOfEnum, listOfListOfCustom, listOfListOfObject);
     }
   }
 }

--- a/apollo-compiler/src/test/graphql/com/example/mutation_create_review/TestOperation.json
+++ b/apollo-compiler/src/test/graphql/com/example/mutation_create_review/TestOperation.json
@@ -198,6 +198,26 @@
 					"type": "Boolean",
 					"description": "for test purpose only",
 					"defaultValue": true
+				},
+				{
+					"name": "listOfListOfString",
+					"type": "[[String]]",
+					"description": "for test purpose only"
+				},
+				{
+					"name": "listOfListOfEnum",
+					"type": "[[Episode]]",
+					"description": "for test purpose only"
+				},
+				{
+					"name": "listOfListOfCustom",
+					"type": "[[Date]]",
+					"description": "for test purpose only"
+				},
+				{
+					"name": "listOfListOfObject",
+					"type": "[[ColorInput]]",
+					"description": "for test purpose only"
 				}
 			]
 		},

--- a/apollo-compiler/src/test/graphql/com/example/mutation_create_review/type/ReviewInput.java
+++ b/apollo-compiler/src/test/graphql/com/example/mutation_create_review/type/ReviewInput.java
@@ -42,11 +42,22 @@ public final class ReviewInput {
 
   private final Input<Boolean> booleanWithDefaultValue;
 
+  private final Input<List<List<String>>> listOfListOfString;
+
+  private final Input<List<List<Episode>>> listOfListOfEnum;
+
+  private final Input<List<List<Object>>> listOfListOfCustom;
+
+  private final Input<List<List<ColorInput>>> listOfListOfObject;
+
   ReviewInput(long stars, Input<Long> nullableIntFieldWithDefaultValue, Input<String> commentary,
       @Nonnull ColorInput favoriteColor, Input<Episode> enumWithDefaultValue,
       Input<Episode> nullableEnum, Input<List<Object>> listOfCustomScalar,
       Input<Object> customScalar, Input<List<Episode>> listOfEnums, Input<List<Long>> listOfInt,
-      Input<List<String>> listOfString, Input<Boolean> booleanWithDefaultValue) {
+      Input<List<String>> listOfString, Input<Boolean> booleanWithDefaultValue,
+      Input<List<List<String>>> listOfListOfString, Input<List<List<Episode>>> listOfListOfEnum,
+      Input<List<List<Object>>> listOfListOfCustom,
+      Input<List<List<ColorInput>>> listOfListOfObject) {
     this.stars = stars;
     this.nullableIntFieldWithDefaultValue = nullableIntFieldWithDefaultValue;
     this.commentary = commentary;
@@ -59,6 +70,10 @@ public final class ReviewInput {
     this.listOfInt = listOfInt;
     this.listOfString = listOfString;
     this.booleanWithDefaultValue = booleanWithDefaultValue;
+    this.listOfListOfString = listOfListOfString;
+    this.listOfListOfEnum = listOfListOfEnum;
+    this.listOfListOfCustom = listOfListOfCustom;
+    this.listOfListOfObject = listOfListOfObject;
   }
 
   /**
@@ -145,6 +160,34 @@ public final class ReviewInput {
     return this.booleanWithDefaultValue.value;
   }
 
+  /**
+   * for test purpose only
+   */
+  public @Nullable List<List<String>> listOfListOfString() {
+    return this.listOfListOfString.value;
+  }
+
+  /**
+   * for test purpose only
+   */
+  public @Nullable List<List<Episode>> listOfListOfEnum() {
+    return this.listOfListOfEnum.value;
+  }
+
+  /**
+   * for test purpose only
+   */
+  public @Nullable List<List<Object>> listOfListOfCustom() {
+    return this.listOfListOfCustom.value;
+  }
+
+  /**
+   * for test purpose only
+   */
+  public @Nullable List<List<ColorInput>> listOfListOfObject() {
+    return this.listOfListOfObject.value;
+  }
+
   public static Builder builder() {
     return new Builder();
   }
@@ -185,7 +228,7 @@ public final class ReviewInput {
             @Override
             public void write(InputFieldWriter.ListItemWriter listItemWriter) throws IOException {
               for (Episode $item : listOfEnums.value) {
-                listItemWriter.writeString($item.name());
+                listItemWriter.writeString($item != null ? $item.name() : null);
               }
             }
           } : null);
@@ -212,6 +255,78 @@ public final class ReviewInput {
         }
         if (booleanWithDefaultValue.defined) {
           writer.writeBoolean("booleanWithDefaultValue", booleanWithDefaultValue.value);
+        }
+        if (listOfListOfString.defined) {
+          writer.writeList("listOfListOfString", listOfListOfString.value != null ? new InputFieldWriter.ListWriter() {
+            @Override
+            public void write(InputFieldWriter.ListItemWriter listItemWriter) throws IOException {
+              for (List<String> $item : listOfListOfString.value) {
+                listItemWriter.writeList($item != null ? new InputFieldWriter.ListWriter() {
+                  @Override
+                  public void write(InputFieldWriter.ListItemWriter listItemWriter) throws
+                      IOException {
+                    for (String $$item : $item) {
+                      listItemWriter.writeString($$item);
+                    }
+                  }
+                } : null);
+              }
+            }
+          } : null);
+        }
+        if (listOfListOfEnum.defined) {
+          writer.writeList("listOfListOfEnum", listOfListOfEnum.value != null ? new InputFieldWriter.ListWriter() {
+            @Override
+            public void write(InputFieldWriter.ListItemWriter listItemWriter) throws IOException {
+              for (List<Episode> $item : listOfListOfEnum.value) {
+                listItemWriter.writeList($item != null ? new InputFieldWriter.ListWriter() {
+                  @Override
+                  public void write(InputFieldWriter.ListItemWriter listItemWriter) throws
+                      IOException {
+                    for (Episode $$item : $item) {
+                      listItemWriter.writeString($$item != null ? $$item.name() : null);
+                    }
+                  }
+                } : null);
+              }
+            }
+          } : null);
+        }
+        if (listOfListOfCustom.defined) {
+          writer.writeList("listOfListOfCustom", listOfListOfCustom.value != null ? new InputFieldWriter.ListWriter() {
+            @Override
+            public void write(InputFieldWriter.ListItemWriter listItemWriter) throws IOException {
+              for (List<Object> $item : listOfListOfCustom.value) {
+                listItemWriter.writeList($item != null ? new InputFieldWriter.ListWriter() {
+                  @Override
+                  public void write(InputFieldWriter.ListItemWriter listItemWriter) throws
+                      IOException {
+                    for (Object $$item : $item) {
+                      listItemWriter.writeCustom(CustomType.DATE, $$item);
+                    }
+                  }
+                } : null);
+              }
+            }
+          } : null);
+        }
+        if (listOfListOfObject.defined) {
+          writer.writeList("listOfListOfObject", listOfListOfObject.value != null ? new InputFieldWriter.ListWriter() {
+            @Override
+            public void write(InputFieldWriter.ListItemWriter listItemWriter) throws IOException {
+              for (List<ColorInput> $item : listOfListOfObject.value) {
+                listItemWriter.writeList($item != null ? new InputFieldWriter.ListWriter() {
+                  @Override
+                  public void write(InputFieldWriter.ListItemWriter listItemWriter) throws
+                      IOException {
+                    for (ColorInput $$item : $item) {
+                      listItemWriter.writeObject($$item != null ? $$item.marshaller() : null);
+                    }
+                  }
+                } : null);
+              }
+            }
+          } : null);
         }
       }
     };
@@ -241,6 +356,14 @@ public final class ReviewInput {
     private Input<List<String>> listOfString = Input.fromNullable(Arrays.<String>asList("test1", "test2", "test3"));
 
     private Input<Boolean> booleanWithDefaultValue = Input.fromNullable(true);
+
+    private Input<List<List<String>>> listOfListOfString = Input.absent();
+
+    private Input<List<List<Episode>>> listOfListOfEnum = Input.absent();
+
+    private Input<List<List<Object>>> listOfListOfCustom = Input.absent();
+
+    private Input<List<List<ColorInput>>> listOfListOfObject = Input.absent();
 
     Builder() {
     }
@@ -344,6 +467,38 @@ public final class ReviewInput {
     /**
      * for test purpose only
      */
+    public Builder listOfListOfString(@Nullable List<List<String>> listOfListOfString) {
+      this.listOfListOfString = Input.fromNullable(listOfListOfString);
+      return this;
+    }
+
+    /**
+     * for test purpose only
+     */
+    public Builder listOfListOfEnum(@Nullable List<List<Episode>> listOfListOfEnum) {
+      this.listOfListOfEnum = Input.fromNullable(listOfListOfEnum);
+      return this;
+    }
+
+    /**
+     * for test purpose only
+     */
+    public Builder listOfListOfCustom(@Nullable List<List<Object>> listOfListOfCustom) {
+      this.listOfListOfCustom = Input.fromNullable(listOfListOfCustom);
+      return this;
+    }
+
+    /**
+     * for test purpose only
+     */
+    public Builder listOfListOfObject(@Nullable List<List<ColorInput>> listOfListOfObject) {
+      this.listOfListOfObject = Input.fromNullable(listOfListOfObject);
+      return this;
+    }
+
+    /**
+     * for test purpose only
+     */
     public Builder nullableIntFieldWithDefaultValueInput(@Nonnull Input<Long> nullableIntFieldWithDefaultValue) {
       this.nullableIntFieldWithDefaultValue = Utils.checkNotNull(nullableIntFieldWithDefaultValue, "nullableIntFieldWithDefaultValue == null");
       return this;
@@ -421,9 +576,41 @@ public final class ReviewInput {
       return this;
     }
 
+    /**
+     * for test purpose only
+     */
+    public Builder listOfListOfStringInput(@Nonnull Input<List<List<String>>> listOfListOfString) {
+      this.listOfListOfString = Utils.checkNotNull(listOfListOfString, "listOfListOfString == null");
+      return this;
+    }
+
+    /**
+     * for test purpose only
+     */
+    public Builder listOfListOfEnumInput(@Nonnull Input<List<List<Episode>>> listOfListOfEnum) {
+      this.listOfListOfEnum = Utils.checkNotNull(listOfListOfEnum, "listOfListOfEnum == null");
+      return this;
+    }
+
+    /**
+     * for test purpose only
+     */
+    public Builder listOfListOfCustomInput(@Nonnull Input<List<List<Object>>> listOfListOfCustom) {
+      this.listOfListOfCustom = Utils.checkNotNull(listOfListOfCustom, "listOfListOfCustom == null");
+      return this;
+    }
+
+    /**
+     * for test purpose only
+     */
+    public Builder listOfListOfObjectInput(@Nonnull Input<List<List<ColorInput>>> listOfListOfObject) {
+      this.listOfListOfObject = Utils.checkNotNull(listOfListOfObject, "listOfListOfObject == null");
+      return this;
+    }
+
     public ReviewInput build() {
       Utils.checkNotNull(favoriteColor, "favoriteColor == null");
-      return new ReviewInput(stars, nullableIntFieldWithDefaultValue, commentary, favoriteColor, enumWithDefaultValue, nullableEnum, listOfCustomScalar, customScalar, listOfEnums, listOfInt, listOfString, booleanWithDefaultValue);
+      return new ReviewInput(stars, nullableIntFieldWithDefaultValue, commentary, favoriteColor, enumWithDefaultValue, nullableEnum, listOfCustomScalar, customScalar, listOfEnums, listOfInt, listOfString, booleanWithDefaultValue, listOfListOfString, listOfListOfEnum, listOfListOfCustom, listOfListOfObject);
     }
   }
 }

--- a/apollo-compiler/src/test/graphql/com/example/mutation_create_review_semantic_naming/TestOperation.json
+++ b/apollo-compiler/src/test/graphql/com/example/mutation_create_review_semantic_naming/TestOperation.json
@@ -198,6 +198,26 @@
 					"type": "Boolean",
 					"description": "for test purpose only",
 					"defaultValue": true
+				},
+				{
+					"name": "listOfListOfString",
+					"type": "[[String]]",
+					"description": "for test purpose only"
+				},
+				{
+					"name": "listOfListOfEnum",
+					"type": "[[Episode]]",
+					"description": "for test purpose only"
+				},
+				{
+					"name": "listOfListOfCustom",
+					"type": "[[Date]]",
+					"description": "for test purpose only"
+				},
+				{
+					"name": "listOfListOfObject",
+					"type": "[[ColorInput]]",
+					"description": "for test purpose only"
 				}
 			]
 		},

--- a/apollo-compiler/src/test/graphql/com/example/mutation_create_review_semantic_naming/type/ReviewInput.java
+++ b/apollo-compiler/src/test/graphql/com/example/mutation_create_review_semantic_naming/type/ReviewInput.java
@@ -42,11 +42,22 @@ public final class ReviewInput {
 
   private final Input<Boolean> booleanWithDefaultValue;
 
+  private final Input<List<List<String>>> listOfListOfString;
+
+  private final Input<List<List<Episode>>> listOfListOfEnum;
+
+  private final Input<List<List<Object>>> listOfListOfCustom;
+
+  private final Input<List<List<ColorInput>>> listOfListOfObject;
+
   ReviewInput(long stars, Input<Long> nullableIntFieldWithDefaultValue, Input<String> commentary,
       @Nonnull ColorInput favoriteColor, Input<Episode> enumWithDefaultValue,
       Input<Episode> nullableEnum, Input<List<Object>> listOfCustomScalar,
       Input<Object> customScalar, Input<List<Episode>> listOfEnums, Input<List<Long>> listOfInt,
-      Input<List<String>> listOfString, Input<Boolean> booleanWithDefaultValue) {
+      Input<List<String>> listOfString, Input<Boolean> booleanWithDefaultValue,
+      Input<List<List<String>>> listOfListOfString, Input<List<List<Episode>>> listOfListOfEnum,
+      Input<List<List<Object>>> listOfListOfCustom,
+      Input<List<List<ColorInput>>> listOfListOfObject) {
     this.stars = stars;
     this.nullableIntFieldWithDefaultValue = nullableIntFieldWithDefaultValue;
     this.commentary = commentary;
@@ -59,6 +70,10 @@ public final class ReviewInput {
     this.listOfInt = listOfInt;
     this.listOfString = listOfString;
     this.booleanWithDefaultValue = booleanWithDefaultValue;
+    this.listOfListOfString = listOfListOfString;
+    this.listOfListOfEnum = listOfListOfEnum;
+    this.listOfListOfCustom = listOfListOfCustom;
+    this.listOfListOfObject = listOfListOfObject;
   }
 
   /**
@@ -145,6 +160,34 @@ public final class ReviewInput {
     return this.booleanWithDefaultValue.value;
   }
 
+  /**
+   * for test purpose only
+   */
+  public @Nullable List<List<String>> listOfListOfString() {
+    return this.listOfListOfString.value;
+  }
+
+  /**
+   * for test purpose only
+   */
+  public @Nullable List<List<Episode>> listOfListOfEnum() {
+    return this.listOfListOfEnum.value;
+  }
+
+  /**
+   * for test purpose only
+   */
+  public @Nullable List<List<Object>> listOfListOfCustom() {
+    return this.listOfListOfCustom.value;
+  }
+
+  /**
+   * for test purpose only
+   */
+  public @Nullable List<List<ColorInput>> listOfListOfObject() {
+    return this.listOfListOfObject.value;
+  }
+
   public static Builder builder() {
     return new Builder();
   }
@@ -185,7 +228,7 @@ public final class ReviewInput {
             @Override
             public void write(InputFieldWriter.ListItemWriter listItemWriter) throws IOException {
               for (Episode $item : listOfEnums.value) {
-                listItemWriter.writeString($item.name());
+                listItemWriter.writeString($item != null ? $item.name() : null);
               }
             }
           } : null);
@@ -212,6 +255,78 @@ public final class ReviewInput {
         }
         if (booleanWithDefaultValue.defined) {
           writer.writeBoolean("booleanWithDefaultValue", booleanWithDefaultValue.value);
+        }
+        if (listOfListOfString.defined) {
+          writer.writeList("listOfListOfString", listOfListOfString.value != null ? new InputFieldWriter.ListWriter() {
+            @Override
+            public void write(InputFieldWriter.ListItemWriter listItemWriter) throws IOException {
+              for (List<String> $item : listOfListOfString.value) {
+                listItemWriter.writeList($item != null ? new InputFieldWriter.ListWriter() {
+                  @Override
+                  public void write(InputFieldWriter.ListItemWriter listItemWriter) throws
+                      IOException {
+                    for (String $$item : $item) {
+                      listItemWriter.writeString($$item);
+                    }
+                  }
+                } : null);
+              }
+            }
+          } : null);
+        }
+        if (listOfListOfEnum.defined) {
+          writer.writeList("listOfListOfEnum", listOfListOfEnum.value != null ? new InputFieldWriter.ListWriter() {
+            @Override
+            public void write(InputFieldWriter.ListItemWriter listItemWriter) throws IOException {
+              for (List<Episode> $item : listOfListOfEnum.value) {
+                listItemWriter.writeList($item != null ? new InputFieldWriter.ListWriter() {
+                  @Override
+                  public void write(InputFieldWriter.ListItemWriter listItemWriter) throws
+                      IOException {
+                    for (Episode $$item : $item) {
+                      listItemWriter.writeString($$item != null ? $$item.name() : null);
+                    }
+                  }
+                } : null);
+              }
+            }
+          } : null);
+        }
+        if (listOfListOfCustom.defined) {
+          writer.writeList("listOfListOfCustom", listOfListOfCustom.value != null ? new InputFieldWriter.ListWriter() {
+            @Override
+            public void write(InputFieldWriter.ListItemWriter listItemWriter) throws IOException {
+              for (List<Object> $item : listOfListOfCustom.value) {
+                listItemWriter.writeList($item != null ? new InputFieldWriter.ListWriter() {
+                  @Override
+                  public void write(InputFieldWriter.ListItemWriter listItemWriter) throws
+                      IOException {
+                    for (Object $$item : $item) {
+                      listItemWriter.writeCustom(CustomType.DATE, $$item);
+                    }
+                  }
+                } : null);
+              }
+            }
+          } : null);
+        }
+        if (listOfListOfObject.defined) {
+          writer.writeList("listOfListOfObject", listOfListOfObject.value != null ? new InputFieldWriter.ListWriter() {
+            @Override
+            public void write(InputFieldWriter.ListItemWriter listItemWriter) throws IOException {
+              for (List<ColorInput> $item : listOfListOfObject.value) {
+                listItemWriter.writeList($item != null ? new InputFieldWriter.ListWriter() {
+                  @Override
+                  public void write(InputFieldWriter.ListItemWriter listItemWriter) throws
+                      IOException {
+                    for (ColorInput $$item : $item) {
+                      listItemWriter.writeObject($$item != null ? $$item.marshaller() : null);
+                    }
+                  }
+                } : null);
+              }
+            }
+          } : null);
         }
       }
     };
@@ -241,6 +356,14 @@ public final class ReviewInput {
     private Input<List<String>> listOfString = Input.fromNullable(Arrays.<String>asList("test1", "test2", "test3"));
 
     private Input<Boolean> booleanWithDefaultValue = Input.fromNullable(true);
+
+    private Input<List<List<String>>> listOfListOfString = Input.absent();
+
+    private Input<List<List<Episode>>> listOfListOfEnum = Input.absent();
+
+    private Input<List<List<Object>>> listOfListOfCustom = Input.absent();
+
+    private Input<List<List<ColorInput>>> listOfListOfObject = Input.absent();
 
     Builder() {
     }
@@ -344,6 +467,38 @@ public final class ReviewInput {
     /**
      * for test purpose only
      */
+    public Builder listOfListOfString(@Nullable List<List<String>> listOfListOfString) {
+      this.listOfListOfString = Input.fromNullable(listOfListOfString);
+      return this;
+    }
+
+    /**
+     * for test purpose only
+     */
+    public Builder listOfListOfEnum(@Nullable List<List<Episode>> listOfListOfEnum) {
+      this.listOfListOfEnum = Input.fromNullable(listOfListOfEnum);
+      return this;
+    }
+
+    /**
+     * for test purpose only
+     */
+    public Builder listOfListOfCustom(@Nullable List<List<Object>> listOfListOfCustom) {
+      this.listOfListOfCustom = Input.fromNullable(listOfListOfCustom);
+      return this;
+    }
+
+    /**
+     * for test purpose only
+     */
+    public Builder listOfListOfObject(@Nullable List<List<ColorInput>> listOfListOfObject) {
+      this.listOfListOfObject = Input.fromNullable(listOfListOfObject);
+      return this;
+    }
+
+    /**
+     * for test purpose only
+     */
     public Builder nullableIntFieldWithDefaultValueInput(@Nonnull Input<Long> nullableIntFieldWithDefaultValue) {
       this.nullableIntFieldWithDefaultValue = Utils.checkNotNull(nullableIntFieldWithDefaultValue, "nullableIntFieldWithDefaultValue == null");
       return this;
@@ -421,9 +576,41 @@ public final class ReviewInput {
       return this;
     }
 
+    /**
+     * for test purpose only
+     */
+    public Builder listOfListOfStringInput(@Nonnull Input<List<List<String>>> listOfListOfString) {
+      this.listOfListOfString = Utils.checkNotNull(listOfListOfString, "listOfListOfString == null");
+      return this;
+    }
+
+    /**
+     * for test purpose only
+     */
+    public Builder listOfListOfEnumInput(@Nonnull Input<List<List<Episode>>> listOfListOfEnum) {
+      this.listOfListOfEnum = Utils.checkNotNull(listOfListOfEnum, "listOfListOfEnum == null");
+      return this;
+    }
+
+    /**
+     * for test purpose only
+     */
+    public Builder listOfListOfCustomInput(@Nonnull Input<List<List<Object>>> listOfListOfCustom) {
+      this.listOfListOfCustom = Utils.checkNotNull(listOfListOfCustom, "listOfListOfCustom == null");
+      return this;
+    }
+
+    /**
+     * for test purpose only
+     */
+    public Builder listOfListOfObjectInput(@Nonnull Input<List<List<ColorInput>>> listOfListOfObject) {
+      this.listOfListOfObject = Utils.checkNotNull(listOfListOfObject, "listOfListOfObject == null");
+      return this;
+    }
+
     public ReviewInput build() {
       Utils.checkNotNull(favoriteColor, "favoriteColor == null");
-      return new ReviewInput(stars, nullableIntFieldWithDefaultValue, commentary, favoriteColor, enumWithDefaultValue, nullableEnum, listOfCustomScalar, customScalar, listOfEnums, listOfInt, listOfString, booleanWithDefaultValue);
+      return new ReviewInput(stars, nullableIntFieldWithDefaultValue, commentary, favoriteColor, enumWithDefaultValue, nullableEnum, listOfCustomScalar, customScalar, listOfEnums, listOfInt, listOfString, booleanWithDefaultValue, listOfListOfString, listOfListOfEnum, listOfListOfCustom, listOfListOfObject);
     }
   }
 }

--- a/apollo-compiler/src/test/graphql/schema.json
+++ b/apollo-compiler/src/test/graphql/schema.json
@@ -1777,6 +1777,96 @@
                 "ofType": null
               },
               "defaultValue": "true"
+            },
+            {
+              "name": "listOfListOfString",
+              "description": "for test purpose only",
+              "args": [],
+              "type": {
+                "kind": "LIST",
+                "name": null,
+                "ofType": {
+                  "kind": "LIST",
+                  "name": null,
+                  "ofType": {
+                    "kind": "SCALAR",
+                    "name": "String",
+                    "ofType": null
+                  }
+                }
+              }
+            },
+            {
+              "name": "listOfListOfEnum",
+              "description": "for test purpose only",
+              "args": [],
+              "type": {
+                "kind": "LIST",
+                "name": null,
+                "ofType": {
+                  "kind": "LIST",
+                  "name": null,
+                  "ofType": {
+                    "kind": "ENUM",
+                    "name": "Episode",
+                    "ofType": null
+                  }
+                }
+              }
+            },
+            {
+              "name": "listOfListOfEnum",
+              "description": "for test purpose only",
+              "args": [],
+              "type": {
+                "kind": "LIST",
+                "name": null,
+                "ofType": {
+                  "kind": "LIST",
+                  "name": null,
+                  "ofType": {
+                    "kind": "ENUM",
+                    "name": "Episode",
+                    "ofType": null
+                  }
+                }
+              }
+            },
+            {
+              "name": "listOfListOfCustom",
+              "description": "for test purpose only",
+              "args": [],
+              "type": {
+                "kind": "LIST",
+                "name": null,
+                "ofType": {
+                  "kind": "LIST",
+                  "name": null,
+                  "ofType": {
+                    "kind": "SCALAR",
+                    "name": "Date",
+                    "ofType": null
+                  }
+                }
+              }
+            },
+            {
+              "name": "listOfListOfObject",
+              "description": "for test purpose only",
+              "args": [],
+              "type": {
+                "kind": "LIST",
+                "name": null,
+                "ofType": {
+                  "kind": "LIST",
+                  "name": null,
+                  "ofType": {
+                    "kind": "INPUT_OBJECT",
+                    "name": "ColorInput",
+                    "ofType": null
+                  }
+                }
+              }
             }
           ],
           "interfaces": null,

--- a/apollo-runtime/src/main/java/com/apollographql/apollo/internal/json/InputFieldJsonWriter.java
+++ b/apollo-runtime/src/main/java/com/apollographql/apollo/internal/json/InputFieldJsonWriter.java
@@ -108,47 +108,69 @@ public class InputFieldJsonWriter implements InputFieldWriter {
     }
 
     @Override public void writeString(String value) throws IOException {
-      if (value != null) {
+      if (value == null) {
+        jsonWriter.nullValue();
+      } else {
         jsonWriter.value(value);
       }
     }
 
     @Override public void writeInt(Integer value) throws IOException {
-      if (value != null) {
+      if (value == null) {
+        jsonWriter.nullValue();
+      } else {
         jsonWriter.value(value);
       }
     }
 
     @Override public void writeLong(Long value) throws IOException {
-      if (value != null) {
+      if (value == null) {
+        jsonWriter.nullValue();
+      } else {
         jsonWriter.value(value);
       }
     }
 
     @Override public void writeDouble(Double value) throws IOException {
-      if (value != null) {
+      if (value == null) {
+        jsonWriter.nullValue();
+      } else {
         jsonWriter.value(value);
       }
     }
 
     @Override public void writeBoolean(Boolean value) throws IOException {
-      if (value != null) {
+      if (value == null) {
+        jsonWriter.nullValue();
+      } else {
         jsonWriter.value(value);
       }
     }
 
     @Override public void writeCustom(ScalarType scalarType, Object value) throws IOException {
-      if (value != null) {
+      if (value == null) {
+        jsonWriter.nullValue();
+      } else {
         CustomTypeAdapter customTypeAdapter = scalarTypeAdapters.adapterFor(scalarType);
         writeString(customTypeAdapter.encode(value));
       }
     }
 
     @Override public void writeObject(InputFieldMarshaller marshaller) throws IOException {
-      if (marshaller != null) {
+      if (marshaller == null) {
+        jsonWriter.nullValue();
+      } else {
         jsonWriter.beginObject();
         marshaller.marshal(new InputFieldJsonWriter(jsonWriter, scalarTypeAdapters));
         jsonWriter.endObject();
+      }
+    }
+
+    @Override public void writeList(ListWriter listWriter) throws IOException {
+      if (listWriter == null) {
+        jsonWriter.nullValue();
+      } else {
+        listWriter.write(new JsonListItemWriter(jsonWriter, scalarTypeAdapters));
       }
     }
   }

--- a/apollo-runtime/src/test/java/com/apollographql/apollo/internal/subscription/SubscriptionManagerTest.java
+++ b/apollo-runtime/src/test/java/com/apollographql/apollo/internal/subscription/SubscriptionManagerTest.java
@@ -92,7 +92,8 @@ public class SubscriptionManagerTest {
 
     assertThat(subscriptionManager.timer.tasks).containsKey(RealSubscriptionManager.INACTIVITY_TIMEOUT_TIMER_TASK_ID);
 
-    subscriptionTransportFactory.subscriptionTransport.disconnectCountDownLatch.awaitOrThrowWithTimeout(RealSubscriptionManager.INACTIVITY_TIMEOUT + 200, TimeUnit.MILLISECONDS);
+    subscriptionTransportFactory.subscriptionTransport.disconnectCountDownLatch.awaitOrThrowWithTimeout
+        (RealSubscriptionManager.INACTIVITY_TIMEOUT + 800, TimeUnit.MILLISECONDS);
     assertThat(subscriptionTransportFactory.subscriptionTransport.disconnectMessage).isInstanceOf(OperationClientMessage.Terminate.class);
     assertThat(subscriptionManager.state).isEqualTo(RealSubscriptionManager.State.DISCONNECTED);
     assertThat(subscriptionManager.timer.tasks).isEmpty();
@@ -104,7 +105,8 @@ public class SubscriptionManagerTest {
     subscriptionTransportFactory.callback.onMessage(new OperationServerMessage.ConnectionAcknowledge());
     subscriptionManager.unsubscribe(subscription1);
 
-    subscriptionTransportFactory.subscriptionTransport.disconnectCountDownLatch.awaitOrThrowWithTimeout(RealSubscriptionManager.INACTIVITY_TIMEOUT + 200, TimeUnit.MILLISECONDS);
+    subscriptionTransportFactory.subscriptionTransport.disconnectCountDownLatch.awaitOrThrowWithTimeout
+        (RealSubscriptionManager.INACTIVITY_TIMEOUT + 800, TimeUnit.MILLISECONDS);
     assertThat(subscriptionManager.state).isEqualTo(RealSubscriptionManager.State.DISCONNECTED);
 
     subscriptionManager.subscribe(subscription2, new SubscriptionManagerCallbackAdapter<Operation.Data>());
@@ -122,7 +124,8 @@ public class SubscriptionManagerTest {
     subscriptionTransportFactory.callback.onConnected();
 
     assertThat(subscriptionManager.timer.tasks).containsKey(RealSubscriptionManager.CONNECTION_ACKNOWLEDGE_TIMEOUT_TIMER_TASK_ID);
-    subscriptionTransportFactory.subscriptionTransport.disconnectCountDownLatch.awaitOrThrowWithTimeout(RealSubscriptionManager.CONNECTION_ACKNOWLEDGE_TIMEOUT + 200, TimeUnit.MILLISECONDS);
+    subscriptionTransportFactory.subscriptionTransport.disconnectCountDownLatch.awaitOrThrowWithTimeout
+        (RealSubscriptionManager.CONNECTION_ACKNOWLEDGE_TIMEOUT + 800, TimeUnit.MILLISECONDS);
     assertThat(subscriptionTransportFactory.subscriptionTransport.disconnectMessage).isInstanceOf(OperationClientMessage.Terminate.class);
     assertThat(subscriptionManager.state).isEqualTo(RealSubscriptionManager.State.DISCONNECTED);
     assertThat(subscriptionManager.timer.tasks).isEmpty();

--- a/gradle/dependencies.gradle
+++ b/gradle/dependencies.gradle
@@ -14,7 +14,7 @@ ext.isCi = "true" == System.getenv('CI')
 
 ext.androidConfig = [
     compileSdkVersion: 25,
-    buildToolsVersion: '25.0.3',
+    buildToolsVersion: '26.0.2',
     minSdkVersion    : 15,
     targetSdkVersion : 25
 ]

